scene/gui/line_edit.cpp

```diff
@@ -1777,6 +1610,7 @@ void LineEdit::_create_undo_state() {
 
 void LineEdit::_generate_context_menu() {
 	// Reorganize context menu.
+	// These cannot use the actions/shortcuts system as it would rely on Editor/Project settings.
```

Comment:
This cannot work in the context of the editor. To get the bindings for these actions, you would need access to EditorSettings, which introduces a dependency on /editor in /scene, which is not desired.

If you access the bindings in ProjectSettings, then you would be getting the events registered for the project, and not for your editor.

Basically: this code might need to get the shortcut from EditorSettings or ProjectSettings... how do you determine which one to use, and then use it, without bringing a dependency on /editor.

Reply:
Ahh I see, would it be possible to go through the `InputMap`? Somthing like: `InputMap::get_singleton()->action_get_events("ui_copy")[0]`?

Reply:
hmmm... I think so. It would have to be cast to InputEventKey first, since an accelerator expects a keycode combination. But also we would need to handle cases where: 1) Physical keycode is used, an 2) there are no events.
```cpp
Ref<InputEventKey> e = action_get_events("thing")[0];
e->get_keycode_with_modifiers();
```

Reply:
Additionally even if we used InputMap to get the event, these shortcuts would not update when the event selected for this shortcut is changed, unless a restart of the application is done.

Reply:
That does make things a little trickier, casting and checking `get_physical_keycode` should be okay for the first points. 

Ideally the user does not have to restart and removing them is a pain for discoverability, even though it makes the most sense, otherwise we'd end up back in the same situation. The only other ways I can think of are:

1) Adding a `action_changed` signal to`InputMap`.
2) Make `_generate_context_menu` virtual and override it for the editor.

Reply:
Number 1 probably seems easier. For number 2,wouldnt that mean we would need a class like EditorLineEdit? And replacing all occurrences of LineEdit used in the editor with that? Seems annoying.

However, couldn't we simplify it still by just calling `generate context menu` right before popping it up?

Reply:
Yeah generating it before popping up sounds good.

Reply:
Here is what it looks like now. Pretty clean if you ask me.

```cpp
int TextEdit::_get_menu_action_accelerator(const String &p_action) {
	const List<Ref<InputEvent>> *events = InputMap::get_singleton()->action_get_events(p_action);
	if (!events) {
		return 0;
	}

	// Use first event in the list for the accelerator.
	const List<Ref<InputEvent>>::Element *first_event = events->front();
	if (!first_event) {
		return 0;
	}

	const Ref<InputEventKey> event = first_event->get();
	if (event.is_null()) {
		return 0;
	}

	// Use physical keycode if non-zero
	if (event->get_physical_keycode() != 0) {
		return event->get_physical_keycode_with_modifiers();
	} else {
		return event->get_keycode_with_modifiers();
	}
}

void TextEdit::_generate_context_menu() {
	// Reorganize context menu.
	menu->clear();
	if (!readonly) {
		menu->add_item(RTR("Cut"), MENU_CUT, is_shortcut_keys_enabled() ? _get_menu_action_accelerator("ui_cut") : 0);
	}
// ... 
```